exercises/src/main/scala/fpinscala/datastructures/List.scala

```diff
@@ -50,19 +50,81 @@ object List { // `List` companion object. Contains functions for creating and wo
     foldRight(ns, 1.0)(_ * _) // `_ * _` is more concise notation for `(x,y) => x * y`; see sidebar
 
 
-  def tail[A](l: List[A]): List[A] = ???
+  def tail[A](l: List[A]): List[A] = l match {
+    case Nil => Nil
+    case Cons(head, tail) => tail
+  }
 
-  def setHead[A](l: List[A], h: A): List[A] = ???
+  def head[A](l: List[A]): A = l match {
+    case Cons(x,xs) => x
+  }
 
-  def drop[A](l: List[A], n: Int): List[A] = ???
+  def setHead[A](l: List[A], h: A): List[A] = Cons(h, l)
 
-  def dropWhile[A](l: List[A], f: A => Boolean): List[A] = ???
+  def drop[A](l: List[A], n: Int): List[A] = n match {
+    case 0 => l
+    case _ => drop(tail(l), n-1)
+  }
 
-  def init[A](l: List[A]): List[A] = ???
+  def dropWhile[A](l: List[A], f: A => Boolean): List[A] = {
+    if (f(head(l))) dropWhile(tail(l), f)
+    else l
+  }
+  // Alternatively,
+  // match f(l.head) {
+  //   case true => dropWhile(l.tail, f)
+  //   case false => l
+  // }
+  
+
+  def init[A](l: List[A]): List[A] = {
+    def rec[A](res: List[A], list: List[A]): List[A] = list match {
+      case Nil => res
+      case Cons(x, y) => rec(Cons(x, res), tail(list))
+    }
+    rec(List(), l)
+  }
+
+  def length[A](l: List[A]): Int = foldRight(l, 0) ( (x, y) => 1 + y )
+
+  // Exercise 3.10
+  def foldLeft[A,B](l: List[A], z: B)(f: (B, A) => B): B = l match {
+    case Nil => z
+    case Cons(x, xs) => foldLeft(xs, f(z,x))(f)
+  }
 
-  def length[A](l: List[A]): Int = ???
+  // Exercise 3.11
+  def sum3(ns: List[Int]) = foldLeft(ns, 0)(_ + _)
+  def prod3(ns: List[Int]) = foldLeft(ns, 1)(_ * _)
+  def length2(ns: List[Int]) = foldLeft(ns, (1-head(ns))) ((x,y)=> x + 1) // ?? maybe this would in fact work with foldRight
```

Comment:
It took me some time to remember exactly how `foldL` and `foldR` are different, but I eventually got it (with the help of some examples/ pen& paper experiments).

I think this line would be a correct implementation of length if it were using foldRight instead